pkg/netdevice/netDeviceProvider.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2020 Intel Corp. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package netdevice
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/golang/glog"
+	"github.com/intel/sriov-network-device-plugin/pkg/types"
+	"github.com/intel/sriov-network-device-plugin/pkg/utils"
+	"github.com/jaypipes/ghw"
+	"github.com/vishvananda/netlink"
+)
+
+type netDeviceProvider struct {
+	deviceList []types.PciNetDevice
+	rFactory   types.ResourceFactory
+}
+
+// NewNetDeviceProvider DeviceProvider implementation from netDeviceProvider instance
+func NewNetDeviceProvider(rf types.ResourceFactory) types.DeviceProvider {
+	return &netDeviceProvider{
+		rFactory:   rf,
+		deviceList: make([]types.PciNetDevice, 0),
+	}
+}
+
+func (np *netDeviceProvider) GetDevices() []types.PciDevice {
+	newPciDevices := make([]types.PciDevice, len(np.deviceList))
+	for i := range np.deviceList {
+		newPciDevices[i] = np.deviceList[i]
+	}
+	return newPciDevices
+}
+
+func (np *netDeviceProvider) AddTargetDevices(devices []*ghw.PCIDevice, deviceCode int) error {
+
+	for _, device := range devices {
+		devClass, err := strconv.ParseInt(device.Class.ID, 16, 64)
+		if err != nil {
+			glog.Warningf("discoverDevices(): unable to parse device class for device %+v %q", device, err)
+			continue
+		}
+
+		if devClass == int64(deviceCode) {
```

Comment:
not sure if there is a use case for FPGA but filtering can be done according to PCI class and subclass.
as PCI spec allows further partitioning of the PCI class into subclasses.

see: https://pcisig.com/sites/default/files/files/PCI_Code-ID_r_1_11__v24_Jan_2019.pdf

e.g for network devices:
The PCI class is 0x02
A subclass of 0x00 represents Ethernet controller and 0x07 represents an Infiniband controllers

Anyway, I reckon that for the time being, using just the class is fine.

Reply:
this can be handled with introducing new selector for device "subclass". But I think the existing "device" selector(which is for device hex code) can already target a very specific device. Not sure if this is needed.